Create CONTRIBUTING.md (#2615)
eclipse/xtext: CONTRIBUTING.md

This page explains common tasks related to working with Xtext's source code.

## Report an Issue

> Please create a GitHub issue in this repository.

Reviewer: Nitpicky: `this repository` might be misleading if you have a local clone / fork. I'd put a link to the new issue page behind `issue`.

Author: Done

> Eclipse Bugzilla issues are deprecated for Xtext. However, we still accept them.

Author: Maybe we might be more direct: ", and we do not accept them anymore"?

> Eclipse Bugzilla issues are deprecated for Xtext. However, we still accept them.

Reviewer 2: foundation will shut it down sooner (or later) anyway

> Eclipse Bugzilla issues are deprecated for Xtext. However, we still accept them.

Reviewer: I'm in favor of making it more concise by removing the reference to bugzilla

### Milestones

> We have one milestone for each Xtext release. Fixed issues should be assigned to the milestone of the nearest release in which the fix will be included. For example, if an issue is fixed on the `maintenance` branch, the corresponding milestone is the next service release (increasing the third version component). If an issue is fixed on the `main` branch, the next major or minor release should be assigned.

Reviewer 3: ```suggestion
We have one milestone for each Xtext release. Fixed issues should be assigned to the milestone of the nearest release in which the fix will be included. For example, if an issue is fixed on the `maintenance` branch, the corresponding milestone is the next service release (increasing the patch version). If an issue is fixed on the `main` branch, the next major or minor release should be assigned.
```

Author: @tivervac I applied a similar suggestion by @szarnekow

> We have one milestone for each Xtext release. Fixed issues should be assigned to the milestone of the nearest release in which the fix will be included. For example, if an issue is fixed on the `maintenance` branch, the corresponding milestone is the next service release (increasing the third version component). If an issue is fixed on the `main` branch, the next major or minor release should be assigned.

Reviewer: ```suggestion
We use a milestone for each Xtext release. Fixed issues should be assigned to the milestone of the upcoming release in which the fix will be included. For example, if an issue is fixed on the `maintenance` branch, the corresponding milestone is the next service release (increasing the patch segment of the version). The next major or minor release should be assigned if an issue is fixed on the `main` branch.
```

It is possible to assign issues to a milestone before they are fixed, which can be useful for planning. However, make sure someone will actually work on that issue when you do that!

> The list of issues assigned to a milestone will give a nice overview of new features and fixed bugs after a release has been done, facilitating the creation of release notes and helping users to know when a fix will be available.

Reviewer 3: ```suggestion
The list of issues assigned to a milestone gives a nice overview of new features and fixed bugs after a release has been done, facilitating the creation of release notes and communicating to users when a fix will be available.
```

### Issue Labels

Issue labels can serve several purposes:

* Indicate the issue type: `bug`, `enhancement`, `new_feature`, `question`
* Communicate the current status of an issue: `confirmed`, `help_wanted`
* Communicate why an issue has not been fixed: `duplicate`, `invalid`, `wontfix`
* Categorize issues for better overview, e.g. to assign an issue to a specific part of the software. Committers may add categorization labels as required. However, in most cases the `[category]` prefixing in the issue title should be sufficient.

> In general we want to keep the number of labels as low as possible in order to avoid contributors to be overwhelmed and to make sure they are actually used. A label is not useful if it's assigned only to a fraction of the issues it belongs to, so it's important for all contributors to use the existing labels consistently.

Reviewer: ```suggestion
In general, we want to keep the number of labels as low as possible to avoid overwhelming contributors and ensure the labels are used. A label is not helpful if it is assigned only to a fraction of the issues it belongs to, and that's why all contributors need to use the existing labels consistently.
```

### Workflow

All committers should feel responsible to read incoming issues. When you read a new issue, please think about an appropriate reaction:

* If possible, assign a type and a status.
* If appropriate, close the issue and explain why it won't be fixed.

> * If the report seems reasonable, a comment with some feedback to the reporter would be nice (e.g. describe wich action is required next, such as to confirm the issue or find a solution for it), especially if the reporter is not a committer.

Reviewer 3: ```suggestion
* If the report seems reasonable, a comment with some feedback to the reporter would be nice (e.g. describe which action is required next to confirm or solve the issue), especially if the reporter is not a committer.
```

* If you are not familiar enough with the topic of the issue, you might find someone who is and delegate the action.

## Set up your Eclipse Workspace
Before starting the setup process, make sure that you have both an [Eclipse](https://accounts.eclipse.org/user/register) and a [GitHub](https://github.com/join) account.

1. Download and start the [Eclipse Installer](https://wiki.eclipse.org/Eclipse_Oomph_Installer).
1. On the initial page, click on the *Switch to Advanced Mode* button in the top right.
1. On the *Product* page, select *Eclipse IDE for Eclipse Committers*.
1. On the *Projects* page, select the *Xtext* node.
1. Choose your preferred installation settings on the *Variables* page. Enter credentials for your Eclipse and GitHub account. If you don't have ssh keys registered with Github then make sure that you select `HTTPS (read-only, anonymous)` or `HTTPS (read-write)` for the Github repository entries.
1. Finish the wizard, drink a cup of coffee, and watch how your Xtext development environment is assembled.

> IMPORTANT: if you had previously set up your workspace when Xtext code was split in several Git repositories, please, start completely from scratch (new installation and new clone).

Reviewer: I'd leave this out. Existing contributors will likely have followed the repo merge. The avg reader of this document doesn't know about it and it would be confusing at best.

Author: Done

> NOTE for Windows users: after the workspace setup has finished, some files in some `xtend-gen` folders will be detected as "dirty" from Git. That is due to some tests using multi-line strings (not multi-line template strings); unfortunately, we still haven't found a solution to this problem. 

Reviewer 3: ```suggestion
NOTE for Windows users: after the workspace setup has finished, some files in some `xtend-gen` folders will be detected as "dirty" by Git. That is due to some tests using multi-line strings (not multi-line template strings); unfortunately, we still haven't found a solution to this problem. 
```

> ## Contribute via Fork

Reviewer 3: ```suggestion
## Contribute via a Fork
```

You need a [GitHub](https://github.com/join) and an [Eclipse](https://accounts.eclipse.org/user/register) account for which you signed the [Eclipse Contributor Agreement](https://accounts.eclipse.org/user/login?destination=user/eca).

 1. Make sure there is a GitHub issue for what you want to work on.

>  2. Announce in the comments section that you want to work on the issue. Also describe the solution you want to implement. To improve the chances for your contribution to be accepted, you'll want to wait for the feedback of the committers.

Reviewer 3: ```suggestion
 2. Announce in the comments section that you want to work on the issue. Also, describe the solution you want to implement. To improve the chance that your contribution is accepted, you'll want to wait for the feedback of the committers.
```

 3. Implement your change. Take care to follow the [quality guidelines](QUALITY_GUIDELINES.md) to improve the chances of your contribution being accepted by a committer.
 4. [Sign off](https://git-scm.com/docs/git-commit#git-commit--s) your commits using the same email address you are using for your Eclipse account.
 5. Run the Maven/Gradle build locally to see if everything compiles and all tests pass.
 6. Push your changes to your forked repository. It is recommended to create a separate branch, this will make it easier to include the feedback from committers and update your changes.
 7. Create a [pull request](https://help.github.com/articles/using-pull-requests/).

> ## Contributing for Committers

Reviewer: Contribute as a Committer

Author: Done

You're a committer if you have write-access to the Xtext git-repositories.

 1. Make sure there is a GitHub issue for what you want to work on
 2. Assign the issue to yourself.
 3. Create a local git branch.
 4. Implement your change. Take care to follow the [quality guidelines](QUALITY_GUIDELINES.md).
 5. [Sign off](https://git-scm.com/docs/git-commit#git-commit--s) your commits using the same email address you are using for your Eclipse account.
 6. Push the branch into the repository.
 7. [Jenkins](https://ci.eclipse.org/xtext) will automatically create a job for your branch and build it. A few GitHub Actions workflows are also started.
 9. If, and only if, all tests have passed, create a pull request and ask somebody who is familiar with the code you modified to review it. Note that pull requests from this repository will not be built: only external pull requests are built, so please make sure your branch is up-to-date with the main branch.
 10. If the reviewer approves, merge.

> ## Create a release locally

Reviewer 3: ```suggestion
## Create a Release Locally
```

 1. Run the Maven build locally by calling the shell script [`full-build.sh`](https://github.com/eclipse/xtext/blob/main/full-build.sh). You may want to pass `-DskipTests` to skip all tests (keep in mind that with tests the build might take more than 1 hour).
 2. Find the artifacts at `build/maven-repository` or `build/p2-repository` in a folder relative to the repository root.

## Consume the Latest Artifacts

> There are two ways / sources: 

Reviewer: ```suggestion
There are two ways/sources: 
```

 * All Maven artifacts are published every 24 hours to the [public Sonatype snapshot repository](https://oss.sonatype.org/content/repositories/snapshots/org/eclipse/xtext/) and can be consumed from there.

>  * The [Jenkins](https://ci.eclipse.org/xtext) archives the created repositories. So you can find the repository of your choice at: `https://ci.eclipse.org/xtext/job/xtext/job/<git-branch>/(lastSuccessfulBuild|<build-number>)/artifact/build/(maven|p2)-repository/`

Reviewer: ```suggestion
 * The [Jenkins](https://ci.eclipse.org/xtext) archives the created repositories. You can find the repository of your choice at: `https://ci.eclipse.org/xtext/job/xtext/job/<git-branch>/(lastSuccessfulBuild|<build-number>)/artifact/build/(maven|p2)-repository/`
```

## Eclipse Contributor Agreement

> Before your contribution can be accepted by the project team contributors must

Reviewer 3: ```suggestion
Before your contribution can be accepted by the project team, contributors must
```
